bitcoin/src/bip32.rs

```diff
@@ -759,19 +832,23 @@ impl Xpub {
     pub fn ckd_pub<C: secp256k1::Verification>(
         &self,
         secp: &Secp256k1<C>,
-        i: ChildNumber,
-    ) -> Result<Xpub, Error> {
-        let (sk, chain_code) = self.ckd_pub_tweak(i)?;
-        let tweaked = self.public_key.add_exp_tweak(secp, &sk.into())?;
+        i: NormalChildNumber,
+    ) -> Xpub {
+        let (sk, chain_code) =
+            self.ckd_pub_tweak(i.into()).expect("resulting key would be invalid");
```

Comment:
Same as above.

Reply:
So `ckd_pub_tweak` can still fail due to https://github.com/rust-bitcoin/rust-bitcoin/blob/584ab34315054f145a8368f81bca2a5cb122639f/bitcoin/src/bip32.rs#L824

hence the `expect` here.

Any suggestions?

Reply:
I think it's mathematically unlikely (same as guessing a private key), but I don't remember certainly, so I'll let @apoelstra to answer this. If it's mathematically unlikely we can except but a comment about likelihood would help.

If it's not unlikely we need to use `Result`.

Reply:
Yes because you're guaranteed to have 32 bytes

https://github.com/rust-bitcoin/rust-bitcoin/blob/584ab34315054f145a8368f81bca2a5cb122639f/bitcoin/src/bip32.rs#L810-L825

and `ckd_pub_tweak` only fails if `SecretKey::from_slice` is 32 bytes within the secp curve order:

https://github.com/rust-bitcoin/rust-bitcoin/blob/584ab34315054f145a8368f81bca2a5cb122639f/bitcoin/src/bip32.rs#L815-L825

That's why I've put the `expect()`, this will not fail. I could use `unwrap()` but that's not recommended inside `src/` code (it's fine for examples and tests though).

Reply:
I mean not on the curve - essentially someone  managing to create a derivation number that tweaks by the inverse of the key. I can't remember if there's hashing or not. If there is it should be impossible.

Reply:
Yep, you're correct on the crypto front.

But why does `ckd_pub_tweak` return a Result if it can't fail. I think any `expect`s should appear next to the HMAC. (If `ckd_pub_tweak` is returning a result because it takes a general child number, then it needs to be inverted with `ckd_pub` -- the fallible methods should call the infallible ones,not the other way around.)

Reply:
Yeah, `ckd_pub_tweak` should just accept `NormalChildNumber` and not return `Result`.